curta/src/chip/hash/sha/keccak256/mod.rs

```diff
@@ -0,0 +1,166 @@
+#[derive(Debug, Clone)]
+pub struct Keccak256PublicData<T> {
+    pub public_w: Vec<U32Value<T>>,
+    pub hash_state: Vec<U32Value<T>>,
+    pub end_bits: Vec<T>,
+}
+
+#[rustfmt::skip]
+pub const KECCAKF_ROTC: [[u32; 5]; 5] = [
+    [0, 1, 62, 28, 27],
+    [36, 44, 6, 55, 20],
+    [3, 10, 43, 25, 39],
+    [41, 45, 15, 21, 8],
+    [18, 2, 61, 56, 14]
+];
+
+#[rustfmt::skip]
+pub const KECCAKF_RNDC: [[u32; 2]; 24] = [
+    [0x00000001, 0x00000000], [0x00008082, 0x00000000],
+    [0x0000808A, 0x80000000], [0x80008000, 0x80000000],
+    [0x0000808B, 0x00000000], [0x80000001, 0x00000000],
+    [0x80008081, 0x80000000], [0x00008009, 0x80000000],
+    [0x0000008A, 0x00000000], [0x00000088, 0x00000000],
+    [0x80008009, 0x00000000], [0x8000000A, 0x00000000],
+    [0x8000808B, 0x00000000], [0x0000008B, 0x80000000],
+    [0x00008089, 0x80000000], [0x00008003, 0x80000000],
+    [0x00008002, 0x80000000], [0x00000080, 0x80000000],
+    [0x0000800A, 0x00000000], [0x8000000A, 0x80000000],
+    [0x80008081, 0x80000000], [0x00008080, 0x80000000],
+    [0x80000001, 0x00000000], [0x80008008, 0x80000000],
+];
+
+// impl constraint for keccack256 
+// state register 
+impl<L: AirParameters> AirBuilder<L> {
+    pub fn keccak_f(&mut self,
+        operations: &mut ByteLookupOperations,
+    )where
+    // SHOULD it be u64instructions?
+    L::Instruction: U32Instructions {
+        // alloate 5*5 u64 register for storing 25 8byte lanes
+        // state is 25 * 64 bit
+        // a[i,j] is (j * 5 + i)th lane
+        //    i
+        // xxxxx
+        // xxxxx
+        // xxxxx
+        // xxxxx  j  3*5 + 3 = 18th lane
+        // xxxxx
+        // i == x axis, refer to column, j == y axis, refer to row, this diff from most competitve programmer's convention, but follow the keccak spec, so pls tolerate. so bit weird but a[i,j] actually refer to j * 5 + i NOT i * 5 + j.
+
+        let round_const = self.alloc_array::<U64Register>(24);
+        // unsure this is correct..
+        // how to constrain it to come constant, i.e. the fixed column for the circuit?
+        for i in 0..24 {
+            // self.assert_equal(&round_const.get(i)., &ByteArrayRegister::from(KECCAKF_RNDC[i]));
+        }
+
+        // maybe need to boundary the initial state?
+        // define boundary constraint for theta
+        let state = self.alloc_array::<U64Register>(25);
+        
+        // theta
+        // how to constrain an array value follows a transition constraint in for loop? 
+        let c_arr = self.alloc_array::<U64Register>(5);
+        let d_arr = self.alloc_array::<U64Register>(5);
+        for x in 0..5 {
+            let mut c_i = state.get(x); 
+            for y in 1..5 {
+                c_i = self.bitwise_xor(&c_i, &state.get(y * 5 + x), operations);
+            }
+            // Does it suffice to constrain for every row of c_i, it need to satisfy its relationship with state at the same row?
+            self.assert_equal_transition(&c_arr.get(x), &c_i);
```

Comment:
note that you're actually being a bit wasteful by allocating columns for both c_arr and the final c_i

Reply:
how to directly enforce a series of transition constraint on c_arr without the helper column?